datumaro/components/annotations/matcher.py

```diff
@@ -0,0 +1,298 @@
+# Copyright (C) 2023 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from typing import Optional
+
+import numpy as np
+from attr import attrib, attrs
+
+from datumaro.components.merge.base import Merger
+from datumaro.util.annotation_util import (
+    OKS,
+    approximate_line,
+    bbox_iou,
+    max_bbox,
+    mean_bbox,
+    segment_iou,
+)
+
+__all__ = [
+    "match_segments",
+    "AnnotationMatcher",
+    "LabelMatcher",
+    "ShapeMatcher",
+    "BboxMatcher",
+    "PolygonMatcher",
+    "MaskMatcher",
+    "PointsMatcher",
+    "LineMatcher",
+    "CaptionsMatcher",
+    "Cuboid3dMatcher",
+    "ImageAnnotationMatcher",
+    "HashKeyMatcher",
+]
+
+
+def match_segments(
+    a_segms,
+    b_segms,
+    distance=segment_iou,
+    dist_thresh=1.0,
+    label_matcher=lambda a, b: a.label == b.label,
+):
+    assert callable(distance), distance
+    assert callable(label_matcher), label_matcher
+
+    a_segms.sort(key=lambda ann: 1 - ann.attributes.get("score", 1))
+    b_segms.sort(key=lambda ann: 1 - ann.attributes.get("score", 1))
+
+    # a_matches: indices of b_segms matched to a bboxes
+    # b_matches: indices of a_segms matched to b bboxes
+    a_matches = -np.ones(len(a_segms), dtype=int)
+    b_matches = -np.ones(len(b_segms), dtype=int)
+
+    distances = np.array([[distance(a, b) for b in b_segms] for a in a_segms])
+
+    # matches: boxes we succeeded to match completely
+    # mispred: boxes we succeeded to match, having label mismatch
+    matches = []
+    mispred = []
+
+    for a_idx, a_segm in enumerate(a_segms):
+        if len(b_segms) == 0:
+            break
+        matched_b = -1
+        max_dist = -1
+        b_indices = np.argsort(
+            [not label_matcher(a_segm, b_segm) for b_segm in b_segms], kind="stable"
+        )  # prioritize those with same label, keep score order
+        for b_idx in b_indices:
+            if 0 <= b_matches[b_idx]:  # assign a_segm with max conf
+                continue
+            d = distances[a_idx, b_idx]
+            if d < dist_thresh or d <= max_dist:
+                continue
+            max_dist = d
+            matched_b = b_idx
+
+        if matched_b < 0:
+            continue
+        a_matches[a_idx] = matched_b
+        b_matches[matched_b] = a_idx
+
+        b_segm = b_segms[matched_b]
+
+        if label_matcher(a_segm, b_segm):
+            matches.append((a_segm, b_segm))
+        else:
+            mispred.append((a_segm, b_segm))
+
+    # *_umatched: boxes of (*) we failed to match
+    a_unmatched = [a_segms[i] for i, m in enumerate(a_matches) if m < 0]
+    b_unmatched = [b_segms[i] for i, m in enumerate(b_matches) if m < 0]
```

Comment:
`b_segms` is not changed in for statement. Do we need `if len(b_segms) == 0: break` in for statement?

Reply:
Fixed by https://github.com/openvinotoolkit/datumaro/pull/901/commits/b3f72d24005794a1d13a5998045a9d0803d6b5a7